### PR TITLE
Fix osemgrep's failure to take absolute paths as scanning roots

### PR DESCRIPTION
**Pull request by mjambon**
The project root for a file outside of git of the form `/a/b` was incorrectly inferred to `.`. It is now correctly inferred as `/a`. This was a problem for commands using process substitution.

Tests are included. A simple command like this one now works:
```
$ semgrep -l python -e hello <(echo hello) --experimental

┌──── ○○○ ────┐
│ Semgrep CLI │
└─────────────┘

  Code scanning at ludicrous speed.

fatal: not a git repository (or any of the parent directories): .git


┌─────────────┐
│ Scan Status │
└─────────────┘
  Scanning 1 file tracked by git with 1 Code rule:
  Scanning 1 file.


┌────────────────┐
│ 1 Code Finding │
└────────────────┘

    /tmp/osemgrep-named-pipe-5264de-63
            1┆ hello



┌──────────────┐
│ Scan Summary │
└──────────────┘
Some files were skipped or only partially analyzed.
  Scan was limited to files tracked by git.

Ran 1 rule on 1 file: 1 finding.
```

Previously, the error was:
```
[00.30][ERROR]: Error: cannot make path "/tmp/osemgrep-named-pipe-29ba07-63" relative to project root "/home/martin/semgrep".
```

### CODE REVIEW

File: libs/commons/UTmp.mli [hunk 1]
```diff
@@ -1,3 +1,5 @@
+(* TODO: decide whether it's 'temp' or 'tmp' and use it consistently. *)
+
```
Comment (mjambon) [inline libs/commons/UTmp.mli]: This was done.

File: libs/commons/UFile.mli [hunk 2]
```diff
@@ -66,6 +66,26 @@ val read_file : ?max_len:int -> Fpath.t -> string
...
+*)
+val with_temp_file :
+  ?contents:string ->
+  ?dir:Fpath.t ->
+  ?persist:bool ->
+  ?prefix:string ->
+  ?suffix:string ->
+  (Fpath.t -> 'a) ->
+  'a
+
```
Comment (mjambon) [inline libs/commons/UFile.mli]: I forgot about this. It should be merged with a similar function in UTmp.

File: libs/commons/UFile.mli [hunk 3]
```diff
@@ -66,6 +66,26 @@ val read_file : ?max_len:int -> Fpath.t -> string
 val with_open_out : Fpath.t -> ((string -> unit) * out_channel -> 'a) -> 'a
 val with_open_in : Fpath.t -> (in_channel -> 'a) -> 'a
 
+(* Create a temporary file, run the user function, and clean up.
+
+   Options:
```
Comment (amchiclet) [inline libs/commons/UFile.mli]: nit: order comments in the same order as parameters

File: libs/git_wrapper/Git_wrapper.mli [hunk 4]
```diff
@@ -97,13 +97,39 @@ type status = {
...
+*)
+val get_project_root_for_file : Fpath.t -> Fpath.t option
+
+(*
+   If the argument is a directory, return the project root associated with
+   the files it contains. If the argument is another kind of file
+   (normally a regular file or a symlink), then the root of the project
+   containing this file is returned.
+*)
+val get_project_root_for_file_or_files_in_dir : Fpath.t -> Fpath.t option
```
Comment (amchiclet) [inline libs/git_wrapper/Git_wrapper.mli:129]: What's the main reason that we don't export only this one function and use it everywhere?

File: src/targeting/Find_targets.ml [hunk 5]
```diff
@@ -97,7 +97,7 @@ type conf = {
   always_select_explicit_targets : bool;
   explicit_targets : Explicit_targets.t;
   (* osemgrep-only: option (see Git_project.ml and the force_root parameter) *)
```
Comment (amchiclet) [inline src/targeting/Find_targets.ml]: This comment may be old. I don't see a "force_root" parameter in Git_project.ml (or let me know if I missed it).

**Review by amchiclet — APPROVED** [review]

**Review by ajbt200128 — APPROVED** [review]
LS changes seem fine

File: libs/commons/UFile.mli (hunk 2 above)
Comment (mjambon) [reply libs/commons/UFile.mli]: Waiting for https://github.com/semgrep/semgrep-proprietary/pull/1447 which adds options to `with_temp_file`.

File: src/targeting/Find_targets.ml (hunk 5 above)
Comment (mjambon) [reply src/targeting/Find_targets.ml]: Yes, you missed. I made the comment more precise.

File: libs/commons/UFile.mli [hunk 6]
```diff
@@ -66,6 +66,26 @@ val read_file : ?max_len:int -> Fpath.t -> string
 val with_open_out : Fpath.t -> ((string -> unit) * out_channel -> 'a) -> 'a
 val with_open_in : Fpath.t -> (in_channel -> 'a) -> 'a
 
+(* Create a temporary file, run the user function, and clean up.
```
Comment (aryx) [inline libs/commons/UFile.mli]: You should use UTmp.ml for those temp related functions, not UFile.ml

**Review by aryx — APPROVED** [review]

File: libs/commons/UFile.mli (hunk 6 above)
Comment (mjambon) [reply libs/commons/UFile.mli]: yep

File: libs/commons/UFile.mli (hunk 2 above)
Comment (mjambon) [reply libs/commons/UFile.mli]: done

**Pull request merged**